terminal.cpp

```diff
@@ -1,11 +1,124 @@
 #include "terminal.h"
 
 #include <cassert>
+#include <cstring>
+#include <fstream>
+#include <map>
+#include <vector>
 
 #include "termnames.h"
 
 namespace gib::terminal {
 
+namespace {
+template <typename T>
+T read_into(std::ifstream *stream) {
+  T val;
+  std::array<char, sizeof(val)> buf;
+  stream->read(buf.data(), buf.size());
+  // TODO endianness
+  ::memcpy(&val, buf.data(), buf.size());
+  return val;
+}
+
+std::string read_until_null(std::ifstream *stream) {
+  std::string s;
+  for (;;) {
+    char c;
+    stream->read(&c, 1);
+    if (c == 0) break;
+    s.append(&c, 1);
+  }
+  return s;
+}
+}  // namespace
+
+TermInfo TermInfo::from_env() {
+  std::string_view term{getenv("TERM")};
+  if (term.empty()) {
+    // fallback mode
+    term = "dumb";
+  }
+
+  std::filesystem::path terminfo_base{"/usr/share/terminfo"};
+  auto terminfo_path{terminfo_base / term.substr(0, 1) / term};
+
+  return TermInfo::from_file(terminfo_path);
+}
+
+std::string_view TermInfo::query_string(const std::string_view &query)
+{
+  return strings_[query];
+}
+
+TermInfo TermInfo::from_file(const std::filesystem::path &path) {
+  std::ifstream terminfo_file{path, std::ios::in | std::ios::binary};
+  enum class InfoType { Legacy, Extended };
+
+  int16_t magic{read_into<int16_t>(&terminfo_file)};
```

Comment:
maybe use auto here and below to avoid specifying the type twice? :thinking: But not important i guess.

Reply:
done